File: go-backend-interview-prep/06_concurrency/04_channel_directions/main.py
"""Channel directions -- Python equivalent of the Go example.

Python has no directional queues. We use naming conventions
(out_q for send, in_q for receive) and documentation to convey intent.
"""

import queue
import threading

SENTINEL = None


def produce(out_q: queue.Queue, count: int) -> None:
    """Send-only by convention: only puts, then signals done."""
    for i in range(1, count + 1):
        out_q.put(i * 10)
    out_q.put(SENTINEL)


def consume(in_q: queue.Queue) -> None:
    """Receive-only by convention: only gets until sentinel."""
    while True:
        val = in_q.get()
        if val is SENTINEL:
            break
        print(f"  received: {val}")


def transform(in_q: queue.Queue, out_q: queue.Queue) -> None:
    """Reads from in_q, doubles, sends to out_q."""
    while True:
        val = in_q.get()
        if val is SENTINEL:
            out_q.put(SENTINEL)
            break
        out_q.put(val * 2)


def generate_nums(lo: int, hi: int) -> queue.Queue:
    """Generator pattern: returns a queue that emits [lo, hi]."""
    q: queue.Queue[int | None] = queue.Queue()

    def _fill():
        for i in range(lo, hi + 1):
            q.put(i)
        q.put(SENTINEL)

    threading.Thread(target=_fill).start()
    return q


def main() -> None:
    # --- Example 1: producer + consumer ---
    print("=== Producer -> Consumer ===")
    ch: queue.Queue = queue.Queue()
    t = threading.Thread(target=produce, args=(ch, 5))
    t.start()
    consume(ch)
    t.join()

    # --- Example 2: pipeline with transform ---
    print("\n=== Producer -> Transform -> Consumer ===")
    raw: queue.Queue = queue.Queue()
    doubled: queue.Queue = queue.Queue()

    t1 = threading.Thread(target=produce, args=(raw, 4))
    t2 = threading.Thread(target=transform, args=(raw, doubled))
    t1.start()
    t2.start()
    consume(doubled)
    t1.join()
    t2.join()

    # --- Example 3: generator pattern ---
    print("\n=== Generator pattern ===")
    nums = generate_nums(1, 5)
    while True:
        n = nums.get()
        if n is SENTINEL:
            break
        print(f"  generated: {n}")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/06_concurrency/04_channel_directions/main.go
package main

import "fmt"

// produce sends values to a send-only channel and closes it.
func produce(out chan<- int, count int) {
	for i := 1; i <= count; i++ {
		out <- i * 10
	}
	close(out) // sender closes
}

// consume reads from a receive-only channel until closed.
func consume(in <-chan int) {
	for val := range in {
		fmt.Printf("  received: %d\n", val)
	}
}

// transform reads from in, doubles values, sends to out.
func transform(in <-chan int, out chan<- int) {
	for val := range in {
		out <- val * 2
	}
	close(out)
}

func main() {
	// --- Example 1: producer + consumer ---
	fmt.Println("=== Producer -> Consumer ===")
	ch := make(chan int) // bidirectional, converts to directional

	go produce(ch, 5) // ch converts to chan<- int
	consume(ch)        // ch converts to <-chan int

	// --- Example 2: pipeline with transform ---
	fmt.Println("\n=== Producer -> Transform -> Consumer ===")
	raw := make(chan int)
	doubled := make(chan int)

	go produce(raw, 4)
	go transform(raw, doubled)
	consume(doubled)

	// --- Example 3: return receive-only channel ---
	fmt.Println("\n=== Generator pattern ===")
	nums := generateNums(1, 5)
	for n := range nums {
		fmt.Printf("  generated: %d\n", n)
	}
}

// generateNums returns a receive-only channel that emits values [lo, hi].
func generateNums(lo, hi int) <-chan int {
	ch := make(chan int)
	go func() {
		for i := lo; i <= hi; i++ {
			ch <- i
		}
		close(ch)
	}()
	return ch
}
